r_code/synthetic_dgp.R

```diff
@@ -0,0 +1,1027 @@
+  # Generate the bounds for the random uniform noise.
+  lim <- switch(shock_name,
+    "day_num" = 0.1,
+    "week_num" = 0.5,
+    1
+  )
+  # Create a map of the distinct shock frequency and the associated shock.
+  f_shocks <- paste(shock_name[1], "shock", sep = "_")
+  shock_map <- date_tib %>%
+    dplyr::select(tidyselect::all_of(shock_name)) %>%
+    dplyr::distinct() %>%
+    dplyr::mutate(!!as.name(f_shocks) := stats::runif(dplyr::n(), -lim, lim))
+  # Map the shocks into the factors following the AR model.
+  tib_out <- date_tib %>%
+    dplyr::inner_join(shock_map, by = shock_name) %>%
+    dplyr::group_by(!!as.name(shock_name)) %>%
+    dplyr::mutate(
+      !!as.name(factor_name) := stats::arima.sim(
+        model = ar_model, n = dplyr::n(),
+        innov = !!as.name(f_shocks) + stats::rnorm(dplyr::n(),
+          sd = ar_sd
+        ),
+        n.start = 500
+      )
+    ) %>%
+    dplyr::ungroup() %>%
+    dplyr::select(tidyselect::all_of(factor_name))
+
+  return(tib_out)
+}
+
+#' Creates extra noisy factors, beyond the base requisiste number.
+#'
+#' @param date_tib Grid of date time, spanning the desired number of periods.
+#' @param factor_name String indicating the desired factor name.
+#' @param ar_model AR model list object for AR(1) process with rho=0.2.
+#' @param ar_sd AR(1) parameter noise term for the factors.
+#'
+#' @return Tibble with the shocks by shock_name, as well as the factors.
+.ExtraFactorHelper <- function(date_tib, factor_name, ar_model, ar_sd) {
+  # Generate a set of unique dates which will be mapped to a shock.
+  date_tib <- date_tib %>%
+    dplyr::group_by(year_num) %>%
+    dplyr::mutate(extra_t = dplyr::row_number()) %>%
+    dplyr::ungroup()
+  unique_dates <- date_tib %>%
+    dplyr::select(time, extra_t) %>%
+    dplyr::distinct(extra_t)
+
+  effective_t <- max(unique_dates)
+  num_shocks <- sample(3:min(25, effective_t), 1) - 1
+
+  shock_map <- unique_dates %>%
+    sample_n(num_shocks) %>%
+    dplyr::mutate(e_shock = stats::runif(dplyr::n(), -1, 1)) %>%
+    dplyr::full_join(unique_dates, by = "extra_t") %>%
+    dplyr::mutate(e_shock = tidyr::replace_na(e_shock, 0)) %>%
+    dplyr::arrange(extra_t)
+
+  # Map the shocks into factors, and structure the output.
+  tib_out <- date_tib %>%
+    dplyr::inner_join(shock_map, by = "extra_t") %>%
+    dplyr::group_by(e_shock) %>%
+    dplyr::mutate(
+      !!as.name(factor_name) := stats::arima.sim(
+        model = ar_model, n = dplyr::n(),
+        innov = e_shock + stats::rnorm(dplyr::n(),
+          sd = ar_sd
+        ),
+        n.start = 500
+      )
+    ) %>%
+    dplyr::ungroup() %>%
+    dplyr::select(tidyselect::all_of(factor_name))
+
+  return(tib_out)
+}
+
+#' Generate time-varying factors.
+#'
+#' @param num_factors_inp Integer number of time-varying, unobserved factors to
+#'    simulate. For freq="monthly", "weekly", "daily", number of factors must be
+#'     at least 3,4,and 5 respectively.
```

Comment:
```suggestion
#'     at least 3, 4, and 5 respectively.
```

Please also update other instances.